### PR TITLE
Experimental allowsRemovalOfLastSavedPaymentMethod

### PR DESCRIPTION
**Pull request by yuki-stripe**
## Summary
See [API Review](https://docs.google.com/document/d/12pqFEANgkDvvwFNWtXOAB_a4vYE0gikWCICj78Su8kQ/edit#heading=h.1kuxwrt88dsj)

## Motivation
https://jira.corp.stripe.com/browse/MOBILESDK-1605

## Testing
Manually tested the cases [here](https://docs.google.com/document/d/12pqFEANgkDvvwFNWtXOAB_a4vYE0gikWCICj78Su8kQ/edit#heading=h.1kuxwrt88dsj)

Automated tests to follow in another PR

## Changelog
Not public.

### CODE REVIEW

File: Example/PaymentSheet Example/PaymentSheet Example/PlaygroundController.swift [hunk 1]
```diff
@@ -107,7 +107,6 @@ class PlaygroundController: ObservableObject {
         if settings.userOverrideCountry != .off {
             configuration.userOverrideCountry = settings.userOverrideCountry.rawValue
         }
```
Comment (wooj-stripe) [inline Example/PaymentSheet Example/PaymentSheet Example/PlaygroundController.swift:109]: Seems like we should update the controller to turn on/off this feature?

File: StripePaymentSheet/StripePaymentSheet/Source/PaymentSheet/Saved Payment Method Screen/SavedPaymentOptionsViewController.swift [hunk 2]
```diff
@@ -422,38 +433,38 @@ extension SavedPaymentOptionsViewController: PaymentOptionCellDelegate {
...
+        else {
+            assertionFailure()
+            return
+        }
+        let viewModel = viewModels[indexPath.row]
+        self.viewModels.remove(at: indexPath.row)
+        // the deletion needs to be in a performBatchUpdates so we make sure it is completed
+        // before potentially leaving edit mode (which triggers a reload that may collide with
+        // this deletion)
+        self.collectionView.performBatchUpdates {
```
Comment (wooj-stripe) [inline StripePaymentSheet/StripePaymentSheet/Source/PaymentSheet/Saved Payment Method Screen/SavedPaymentOptionsViewController.swift:447]: I'm seeing a problem where the pay button is disabled after deleting a payment method.
1. set `allowRemovalOfLastSavedPaymentMethod == false`
2.  add two payment methods (PM1, and PM2)
3. present payment sheet, tap edit
4. Delete PM2

Observed:
PM2 is deleted, edit mode is disabled, PM1 is selected, but the pay button is greyed out.

File: StripePaymentSheet/StripePaymentSheet/Source/PaymentSheet/CustomerSheet/CustomerSavedPaymentMethodsCollectionViewController.swift [hunk 3]
```diff
@@ -424,6 +424,7 @@ extension CustomerSavedPaymentMethodsCollectionViewController: PaymentOptionCell
                                               removeSavedPaymentMethodMessage: savedPaymentMethodsConfiguration.removeSavedPaymentMethodMessage,
                                               appearance: appearance,
                                               hostedSurface: .customerSheet,
+                                              canRemoveCard: true,
```
Comment (wooj-stripe) [inline StripePaymentSheet/StripePaymentSheet/Source/PaymentSheet/CustomerSheet/CustomerSavedPaymentMethodsCollectionViewController.swift:427]: I'm guessing this is unfinished, and we'll eventually pass in a configuration here?

File: StripePaymentSheet/StripePaymentSheet/Source/PaymentSheet/Saved Payment Method Screen/SavedPaymentOptionsViewController.swift [hunk 4]
```diff
@@ -71,13 +71,24 @@ class SavedPaymentOptionsViewController: UIViewController {
...
-        )
+    // MARK: - Internal Properties
+
+    /// Whether or not you can edit save payment methods by removing or updating them.
+    var canEditPaymentMethods: Bool {
+        switch savedPaymentMethods.count {
+        case 0:
+            return false
+        case 1:
+            // If there's exactly one PM, customer can only edit if configuration allows removal or if that single PM is editable
```
Comment (wooj-stripe) [inline StripePaymentSheet/StripePaymentSheet/Source/PaymentSheet/Saved Payment Method Screen/SavedPaymentOptionsViewController.swift:85]: maybe replace "is editable"
with "allows for the card brand choice to be updated."

File: StripePaymentSheet/StripePaymentSheet/Source/PaymentSheet/Saved Payment Method Screen/SavedPaymentOptionsViewController.swift (hunk 2 above)
Comment (yuki-stripe) [reply StripePaymentSheet/StripePaymentSheet/Source/PaymentSheet/Saved Payment Method Screen/SavedPaymentOptionsViewController.swift:447]: Ah thanks for finding, looks like this is an extension of an existing bug. I have the fix but will put it in a separate follow-up PR.

File: StripePaymentSheet/StripePaymentSheet/Source/PaymentSheet/CustomerSheet/CustomerSavedPaymentMethodsCollectionViewController.swift (hunk 3 above)
Comment (yuki-stripe) [reply StripePaymentSheet/StripePaymentSheet/Source/PaymentSheet/CustomerSheet/CustomerSavedPaymentMethodsCollectionViewController.swift:427]: Yep!

File: Example/PaymentSheet Example/PaymentSheet Example/PlaygroundController.swift (hunk 1 above)
Comment (yuki-stripe) [reply Example/PaymentSheet Example/PaymentSheet Example/PlaygroundController.swift:109]: Will do this in another PR that adds UI tests.  Shame to take up real estate for such a small and _almost_ unused property though, our test playground settings UI isn't scaling so well these days.

File: StripePaymentSheet/StripePaymentSheet/Source/PaymentSheet/Saved Payment Method Screen/SavedPaymentOptionsViewController.swift (hunk 2 above)
Comment (yuki-stripe) [reply StripePaymentSheet/StripePaymentSheet/Source/PaymentSheet/Saved Payment Method Screen/SavedPaymentOptionsViewController.swift:447]: https://github.com/stripe/stripe-ios/pull/3258

**Review by wooj-stripe — APPROVED** [review]

File: StripePaymentSheet/StripePaymentSheet/Source/PaymentSheet/PaymentSheetConfiguration.swift [hunk 5]
```diff
@@ -195,7 +195,7 @@ extension PaymentSheet {
         /// This is an experimental feature that may be removed at any time.
         /// If true (the default), the customer can delete all saved payment methods.
         /// If false, the customer can't delete if they only have one saved payment method remaining.
-        @_spi(ExperimentalAllowRemovalOfLastSavedPaymentMethodAPI) public var allowRemovalOfLastSavedPaymentMethod = true
+        @_spi(ExperimentalAllowsRemovalOfLastSavedPaymentMethodAPI) public var allowsRemovalOfLastSavedPaymentMethod = true
```
Comment (yuki-stripe) [inline StripePaymentSheet/StripePaymentSheet/Source/PaymentSheet/PaymentSheetConfiguration.swift:198]: https://docs.google.com/document/d/12pqFEANgkDvvwFNWtXOAB_a4vYE0gikWCICj78Su8kQ/edit#heading=h.x1tvwe1macqz

**Review by wooj-stripe — APPROVED** [review]

**Pull request merged**